### DISCUSSION TITLE
Berry - how to subscribe to MQTT topic?

### DISCUSSION THREAD

**Discussion by PhilipTS** [arendst/Tasmota#14403] [Q&A] [answered]
Is there a way to subscribe to a MQTT topic from berry?
I am now doing it from the old rules like ... (short version)

Rule1
ON mqtt#connected  DO event subscribe ENDON
ON Event#subscribe DO Subscribe target, %topic%/set/targetTemperature  ENDON
ON Event#target DO TuyaSend2 16,%value% ENDON

**Comment by barbudor** [comment]
You can still use `tasmota.cmd("Subscribe XXX, .....")
Any MQTT message received by Tasmota will be distributed to `mqtt_data()` driver's callback
```
class mqttTest
  def mqtt_data(topic, idx, payload_s, payload_b)
    .... process 
    return false
  end
end
tasmota.add_driver(mqttTest())
```
The function should return `true` if it has handled the message (and prevent further Tasmota modules to look into it), or `false` if the message is not for this driver and need to be forwarded to other Tasmota modules.

**Comment by PhilipTS** [comment]
with the old rule engine the command has this format ...
Subscribe eventName, mqttTopic

So in ```tasmota.cmd("Subscribe XXX, .....")``` i replace the XXX with a function name instead of an event?

did some test ... not sure i get it ... it's not working

````
def mqtt_data(topic, idx, payload_s, payload_b)
  print(topic)
  return(false)
end

tasmota.cmd("Subscribe mqtt_data, tasmota/set/testVar")
````

**Reply by barbudor** [reply]
You keep the same syntax, the `...` just mean I was too lazy to copy the parameters
Event name doesn't care as Berry's `mqtt_data()` will receive every MQTT messages even those sent to `cmnd/xxx/cmnd`
So you have to parse the message and decide if the message is for you (and return True) or not for you (and return false)

**Reply by sfromis** [reply]
If you just want to implement a command to set something, don't bother with MQTT subscribe, in Berry it is much simpler to register a command, which will be called when detected by Tasmota. See `tasmota.add_cmd`. Then it will also work in the console, and other avenues for issuing commands.

**Reply by PhilipTS** [reply]
@sfromis I need to set from MQTT. Yesterday I was trying to figure out Hex -> Int and the bytes() command was perfect.
Today I figured out the last part that I had in the old rules engine, the subscribe part.
Will post the code when I finish for other to see, as an example.
TuyaMCU Thermostat (ME102H.31) in 100% Tasmota Berry.

Thank You Both

**Comment by PhilipTS** [comment, marked as answer]
OK so I have to register it as a driver. Did some reading about Drivers and I got it.

Thank You :)

````
class mqttTest
  def mqtt_data(topic, idx, payload_s, payload_b)
    print(topic)
    return(false)
  end
end
tasmota.add_driver(mqttTest())
tasmota.cmd("Subscribe dummyEvent, tasmota/set/testVar")
````

**Reply by SvenMb** [reply]
in current tasmota this doesn't work anymore, when I subscribe a mqtt topic like this, the drivers `mqtt_data()` function won't get the mqtt message anymore. (I think this is true since tasmota 11)

you should now use `mqtt.subscribe()`, which calls a user defined function only when topic matches.

also if you still like to use drivers, they will get these subscribed topic too.

see: https://tasmota.github.io/docs/Berry/#mqtt-module

**Reply by sfromis** [reply]
While the newer method you link is much enhanced, and indeed strongly recommended now, the older example should still work, maybe depending on circumstances.

(And for being able to set data via MQTT, registering a command is still recommended over doing extra MQTT subscriptions)

**Reply by SvenMb** [reply]
even if you use `mqtt_data()` inside a driver on current tasmota, you will only get data you have subscribed for (at least that is my experience).

This is on par with current documentation:
~~~
mqtt_data(topic, idx, data, databytes): called for MQTT payloads matching mqtt.subscribe. idx is zero, and data is normally unparsed JSON.
~~~

maybe it is depending on a `import mqtt`, but I would not count on it.

**Reply by s-hadinger** [reply]
Right. That was a change made to not flood berry drivers with messages they are not interested in. Performance might suffer otherwise
